ENH FEA add interaction constraints to HGBT

lorentzenchr opened the pull request:
#### Reference Issues/PRs
Fixes #19148.

#### What does this implement/fix? Explain your changes.
This PR introduces the argument `interaction_cst` to the classes `HistGradientBoostingRegressor` and `HistGradientBoostingClassifier`.

It main impact is in `Splitter.find_node_split`.

#### Any other comments?
Might be a good reason to wish for an early 1.1 release :smirk:

- [x] add new feature
- [x] validate `interaction_cst`
- [x] test new feature
- [x] optimize code
- [x] doc and whatsnew
- [x] test interaction delta of gradient boosting, idea `hgbt(X + X[:, 0]) - hgbt(X)` with and without constraints, see https://github.com/scikit-learn/scikit-learn/pull/21020/commits/ba78cb9db2439e147accdadb702b7bc73354a775
- [x] add to an example, e.g. partial dependence plot should be parallel with constraints and in link space

lorentzenchr committed:
DOC add attribues to TreeGrower

lorentzenchr committed:
ENH add interaction_cst

lorentzenchr committed:
use a set in _get_allowed_features

lorentzenchr committed:
complete overhaul

lorentzenchr commented:
@thomasjpfan I have two questions right now:
1. How to treat features not listed in the interaction constraints, e.g. with features 0..3 and `interaction_cst = [{0, 1}]`, are features 2 and 3 allowed to enter the game at all?
2. What would be a good test for `test_splitting`? I might need a little help there.

lorentzenchr committed:
TST test_split_interaction_constraints

lorentzenchr committed:
DOC add is_leaf to Attributes section

lorentzenchr committed:
DOC improve interaction_cst_idx

lorentzenchr committed:
FIX fix logic

lorentzenchr committed:
TST add test_grower_interaction_constraints

lorentzenchr commented:
Now, I could use some help as I'm stuck with the following error, when compiled with `# cython: boundscheck=True` in `splitting.pyx`:
```
sklearn/ensemble/_hist_gradient_boosting/tests/test_grower.py:247: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
sklearn/ensemble/_hist_gradient_boosting/grower.py:395: in grow
    self.split_next()
        self       = <sklearn.ensemble._hist_gradient_boosting.grower.TreeGrower object at 0x7fe5d39d55d0>
sklearn/ensemble/_hist_gradient_boosting/grower.py:500: in split_next
    ) = self.splitter.split_indices(node.split_info, node.sample_indices)
        node       = <sklearn.ensemble._hist_gradient_boosting.grower.TreeNode object at 0x7fe5d39d5f50>
        self       = <sklearn.ensemble._hist_gradient_boosting.grower.TreeGrower object at 0x7fe5d39d55d0>
        tic        = 1.545465422
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   &sample_indices[right_offset[thread_idx]]
E   IndexError: Out of bounds on buffer access (axis 0)
```
I don't understand where any of `sample_indices`, `right_offset` and `thread_idx` have been changed by this PR.

ogrisel commented:
> Now, I could use some help as I'm stuck with the following error, which only happens for n_threads > 1 (this error message is produced by compiling with # cython: boundscheck=True in splitting.pyx).

I don't understand how a build-time error can be impacted by a runtime value (`n_threads > 1`)...

lorentzenchr commented:
> > Now, I could use some help as I'm stuck with the following error, which only happens for n_threads > 1 (this error message is produced by compiling with # cython: boundscheck=True in splitting.pyx).
> 
> I don't understand how a build-time error can be impacted by a runtime value (`n_threads > 1`)...

Me neither. There is nothing I understand with this error.

Running `test_grower_interaction_constraints` with `TreeGrower(..., n_threads=1)` is fine, while `TreeGrower(..., n_threads=2)` throws this error. It seems as if the calculation of `right_offset` is changed by this PR and now wrong. Some print out gives me (`print(..., flush=True)` is helpful in this case)
```
thread_idx = 1
sample_indices = [5 5]
right_offset = [1 2]
right_indices_buffer = [         1          4 3215170148 1069022147          6          9
 1036367913 1068865250 1066764793 1064467143]
offset_in_buffers = [0 1]
right_counts = [1 0]
len(sample_indices) = 2
right_offset[thread_idx] = 2
```
Note the last two lines and the fact that `sample_indices[right_offset[thread_idx]] = sample_indices[2]` gets called. But why does any of this PR change those calculations in the first place. I'm deeply puzzled.

lorentzenchr committed:
CLN make allowed_features an instance variable

lorentzenchr committed:
TST restructure  test_grower_interaction_constraints

lorentzenchr committed:
CLN improve logic

lorentzenchr committed:
TST improve test

lorentzenchr commented on sklearn/ensemble/_hist_gradient_boosting/splitting.pyx:
```diff
@@ -388,11 +394,12 @@ cdef class Splitter:
-                    &right_indices_buffer[offset_in_buffers[thread_idx]],
-                    sizeof(unsigned int) * right_counts[thread_idx]
-                )
+                if right_counts[thread_idx] > 0:
+                    memcpy(
+                        &sample_indices[right_offset[thread_idx]],
+                        &right_indices_buffer[offset_in_buffers[thread_idx]],
+                        sizeof(unsigned int) * right_counts[thread_idx]
+                    )
```
This was crucial for fixing the out of bound error, but I don't know why this is necessary only with adding this PR while it was not necessary before.

thomasjpfan replied:
Are you referencing the issue with `test_grower_interaction_constraints` mentioned in https://github.com/scikit-learn/scikit-learn/pull/21020#issuecomment-921158509 ? 

(I am having trouble reproducing the out of bound error removing `if` condition.)

jjerphan replied:
I can reproduce the `IndexError` at running time (when running the given test as reported in https://github.com/scikit-learn/scikit-learn/pull/21020#issuecomment-922337842) after having compiled with `boundscheck=True`. I haven't been able to identify the cause of this error.

Fortunately, compiling without bounds checks on memory-views (`boundscheck=False`) works and all HGBT tests pass even without this `if`.

lorentzenchr replied:
@jjerphan Wanna have a look at #21130 where this is fixed and described properly?

jjerphan replied:
I will have a look at it this afternoon. :+1:

lorentzenchr committed:
DOC add docstring for interaction_cst

lorentzenchr committed:
ENH add validation of interaction_cst

lorentzenchr committed:
TST test input validation

lorentzenchr committed:
DEBUG uncomment if condition

lorentzenchr commented:
@thomasjpfan @ogrisel Here comes a detailed bug report based on commit https://github.com/scikit-learn/scikit-learn/pull/21020/commits/3ea38292497d5072f346f0248244c4c2ec1fc985.
```
pytest -xl sklearn/ensemble/_hist_gradient_boosting/tests/test_grower.py::test_grower_interaction_constraints 
======================================================================== test session starts ========================================================================
platform darwin -- Python 3.7.9, pytest-6.2.1, py-1.10.0, pluggy-0.13.1
rootdir: XXX/scikit-learn, configfile: setup.cfg
plugins: anyio-3.3.0, cov-2.10.1
collected 1 item                                                                                                                                                    

sklearn/ensemble/_hist_gradient_boosting/tests/test_grower.py F                                                                                               [100%]

============================================================================= FAILURES ==============================================================================
________________________________________________________________ test_grower_interaction_constraints ________________________________________________________________

    def test_grower_interaction_constraints():
        """Check that grower respects interaction constraints."""
        n_features = 6
        interaction_cst = [{0, 1}, {1, 2}, {3, 4, 5}]
        n_samples = 5
        n_bins = 6
        root_feature_splits = []
    
        def get_all_children(node):
            res = []
            if node.is_leaf:
                return res
            for n in [node.left_child, node.right_child]:
                res.append(n)
                if not n.is_leaf:
                    res.extend(get_all_children(n))
            return res
    
        for seed in range(20):
            rng = np.random.RandomState(seed)
    
            X_binned = rng.randint(
                0, n_bins - 1, size=(n_samples, n_features), dtype=X_BINNED_DTYPE
            )
            X_binned = np.asfortranarray(X_binned)
            gradients = rng.normal(size=n_samples).astype(G_H_DTYPE)
            hessians = np.ones(shape=1, dtype=G_H_DTYPE)
    
            grower = TreeGrower(
                X_binned,
                gradients,
                hessians,
                max_depth=3,
                n_bins=n_bins,
                shrinkage=1.0,
                max_leaf_nodes=None,
                min_samples_leaf=1,
                interaction_cst=interaction_cst,
                n_threads=n_threads,
            )
>           grower.grow()

X_binned   = array([[4, 2, 4, 2, 4, 0],
       [4, 1, 0, 3, 3, 2],
       [3, 2, 2, 3, 3, 3],
       [0, 1, 4, 4, 4, 3],
       [1, 1, 1, 2, 2, 4]], dtype=uint8)
get_all_children = <function test_grower_interaction_constraints.<locals>.get_all_children at 0x7ff910990950>
gradients  = array([-0.29139397, -0.13309029, -0.1730696 , -1.7616516 , -0.08767307],
      dtype=float32)
grower     = <sklearn.ensemble._hist_gradient_boosting.grower.TreeGrower object at 0x7ff910945550>
hessians   = array([1.], dtype=float32)
interaction_cst = [{0, 1}, {1, 2}, {3, 4, 5}]
n_bins     = 6
n_features = 6
n_samples  = 5
rng        = RandomState(MT19937) at 0x7FF9108D76B0
root_feature_splits = []
seed       = 0

sklearn/ensemble/_hist_gradient_boosting/tests/test_grower.py:612: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
sklearn/ensemble/_hist_gradient_boosting/grower.py:397: in grow
    self.split_next()
        self       = <sklearn.ensemble._hist_gradient_boosting.grower.TreeGrower object at 0x7ff910945550>
sklearn/ensemble/_hist_gradient_boosting/grower.py:497: in split_next
    ) = self.splitter.split_indices(node.split_info, node.sample_indices)
        node       = <sklearn.ensemble._hist_gradient_boosting.grower.TreeNode object at 0x7ff9109c0c10>
        self       = <sklearn.ensemble._hist_gradient_boosting.grower.TreeGrower object at 0x7ff910945550>
        tic        = 1.203303851
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   &left_indices_buffer[offset_in_buffers[thread_idx]],
E   IndexError: Out of bounds on buffer access (axis 0)

HISTOGRAM_DTYPE = dtype([('sum_gradients', '<f8'), ('sum_hessians', '<f8'), ('count', '<u4')])
SplitInfo  = <class 'sklearn.ensemble._hist_gradient_boosting.splitting.SplitInfo'>
Splitter   = <class 'sklearn.ensemble._hist_gradient_boosting.splitting.Splitter'>
__builtins__ = <builtins>
__doc__    = 'This module contains routines and data structures to:\n\n- Find the best possible split of a node. For a given node, ... split to a node, i.e. split the indices of the samples at the node\n  into the newly created left and right childs.\n'
__file__   = 'XXX/scikit-learn/sklearn/ensemble/_hist_gradient_boosting/splitting.cpython-37m-darwin.so'
__loader__ = <_frozen_importlib_external.ExtensionFileLoader object at 0x7ff91093bd10>
__name__   = 'sklearn.ensemble._hist_gradient_boosting.splitting'
__package__ = 'sklearn.ensemble._hist_gradient_boosting'
__pyx_unpickle_Enum = <built-in function __pyx_unpickle_Enum>
__pyx_unpickle_Splitter = <built-in function __pyx_unpickle_Splitter>
__spec__   = ModuleSpec(name='sklearn.ensemble._hist_gradient_boosting.splitting', loader=<_frozen_importlib_external.ExtensionFile...origin='XXX/scikit-learn/sklearn/ensemble/_hist_gradient_boosting/splitting.cpython-37m-darwin.so')
__test__   = {}
compute_node_value = <built-in function compute_node_value>
np         = <module 'numpy' from 'XXX/python3_sklearn/lib/python3.7/site-packages/numpy/__init__.py'>

sklearn/ensemble/_hist_gradient_boosting/splitting.pyx:394: IndexError
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! stopping after 1 failures !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
=================================================================== 1 failed, 1 warning in 0.35s ====================================================================

```

System:
    python: 3.7.9 (v3.7.9:13c94747c7, Aug 15 2020, 01:31:08)  [Clang 6.0 (clang-600.0.57)]
executable: XXX/python3_sklearn/bin/python3.7
   machine: Darwin-20.5.0-x86_64-i386-64bit

Python dependencies:
          pip: 21.2.4
   setuptools: 51.3.1
      sklearn: 1.1.dev0
        numpy: 1.20.3
        scipy: 1.6.3
       Cython: 0.29.21
       pandas: 1.2.0
   matplotlib: 3.3.3
       joblib: 1.0.0
threadpoolctl: 2.1.0

Built with OpenMP: True

lorentzenchr commented:
@mayer79 ping

NicolasHug commented on sklearn/ensemble/_hist_gradient_boosting/splitting.pyx:
```diff
@@ -504,12 +528,13 @@ cdef class Splitter:
-            best_feature_idx = self._find_best_feature_to_split_helper(
-                split_infos)
-            split_info = split_infos[best_feature_idx]
+            # split_info is the index of the best split_info
```
I think this should be 

```suggestion
            # split_info is the best split_info
```

as it's not an index, but I'm not sure this comment is needed?

NicolasHug commented on sklearn/ensemble/_hist_gradient_boosting/gradient_boosting.py:
```diff
@@ -1093,6 +1126,16 @@ class HistGradientBoostingRegressor(RegressorMixin, BaseHistGradientBoosting):
+        Each inner list or set specifies the feature indices that are allowed
+        to interact with each other, meaning splits in child nodes are only
+        allowed in those sets. For instance, `[{0, 1}, {2, 3, 4}]` says that
+        a single tree is either only split on features 0 and 1 or only on
```
If I understand the code correctly, the constrains don't constrain the whole tree, they constrain single branches? I would just change into 

> For instance, `[{0, 1}, {2, 3, 4}]` says that a given branch of a tree is either only split...

We can indicate that a branch is a path from the root the a leave, but I'm not sure it's needed

NicolasHug commented on sklearn/ensemble/_hist_gradient_boosting/tests/test_splitting.py:
```diff
@@ -856,3 +856,70 @@ def test_splitting_categorical_sanity(
+        allowed_features=allowed_features,
+    )
+
+    # only features 0 and 1 are allowed to be split on
```
```suggestion
    # only features 0 and 3 are allowed to be split on
```

NicolasHug commented on sklearn/ensemble/_hist_gradient_boosting/tests/test_grower.py:
```diff
@@ -567,3 +567,73 @@ def test_ohe_equivalence(min_samples_leaf, n_unique_categories, target):
+            X_binned,
+            gradients,
+            hessians,
+            max_depth=3,
```
max depth is 3, should we allow the tree to be deeper to give a better change for potential bugs to be caught?

lorentzenchr replied:
` max_depth=5` and `n_samples=10`, instead of 3 and 5.

NicolasHug commented on sklearn/ensemble/_hist_gradient_boosting/gradient_boosting.py:
```diff
@@ -1093,6 +1126,16 @@ class HistGradientBoostingRegressor(RegressorMixin, BaseHistGradientBoosting):
 
         .. versionadded:: 0.23
 
+    interaction_cst : list of lists or sets of int, default=None
```
Could we just say "list of iterable" or "list of array-like" or something like that? And same in the error messages.

Seems a bit unnecessary to special-case for sets?

lorentzenchr replied:
`iterable of iterables of int, default=None`

NicolasHug commented on sklearn/ensemble/_hist_gradient_boosting/tests/test_splitting.py:
```diff
@@ -856,3 +856,70 @@ def test_splitting_categorical_sanity(
+
+def test_split_interaction_constraints():
+    """Check that allowed_features are respected."""
+    rng = np.random.RandomState(919)
```
should we parametrize this over multiple seeds if it doesn't take too long?
Alternatively, I would suggest to handcraft a scenario where a non-constrained tree would pick either feature 1 and 2, and to assert that in the test, to show that it is indeed the interaction constrain and not just pure luck that leads to a feature in {0, 3}

lorentzenchr replied:
Better now with 6a0205840816525fa89bfe1a6b19527c1316a7f8?

NicolasHug commented on sklearn/ensemble/_hist_gradient_boosting/gradient_boosting.py:
```diff
@@ -1405,6 +1450,16 @@ class HistGradientBoostingClassifier(ClassifierMixin, BaseHistGradientBoosting):
+        to interact with each other, meaning splits in child nodes are only
+        allowed in those sets. For instance, `[{0, 1}, {2, 3, 4}]` says that
+        a single tree is either only split on features 0 and 1 or only on
+        features 2, 3 and 4. If there are more featres, e.g. 5 and 6, those
```
```suggestion
        features 2, 3 and 4. If there are more features, e.g. 5 and 6, those
```

lorentzenchr replied:
I'll change this behaviour, i.e. not specified features will automatically go in as one group.

NicolasHug commented on sklearn/ensemble/_hist_gradient_boosting/grower.py:
```diff
@@ -580,6 +637,42 @@ def split_next(self):
+        -------
+        allowed_features : None or ndarray, dtype=int
+            Indices of features allowed to split for children.
+        interaction_cst_idx : None or list of ints
```
This should probably be `interaction_cst_indices` to follow our conventions? e.g. `sample_indices`, `categorical_features_indices`, etc.

NicolasHug reviewed (commented):
Thanks @lorentzenchr , I made  a quick pass with minor comments, this looks great!

NicolasHug commented:
@lorentzenchr regarding the bound check error, do you observe the same error on the `main` branch with `cython: boundscheck=True`? If yes, it's probably not related to this PR. I deactivated the boundchecks for the code to be as fast as possible, there should be no need to put it back IMHO.

> How to treat features not listed in the interaction constraints, e.g. with features 0..3 and interaction_cst = [{0, 1}], are features 2 and 3 allowed to enter the game at all?

I see 3 options

- don't allow them anywhere  (doesn't sound reasonable, users could just remove these features)
- allow them in any branch
- make them a constraint group of their own as suggested in https://github.com/scikit-learn/scikit-learn/issues/19148#issuecomment-923264153.

What is LGBM or XGBoost choice for this?

lorentzenchr commented:
Concerning the `boundscheck`. The out of bound error does not appear on master. The error started to occur with the changed for-loop in `find_node_split` from `for feature_idx in prange(n_features,..)` to `for split_info_idx in prange(n_allowed_features,..)`.

lorentzenchr committed:
address review comments

lorentzenchr committed:
Revert "DEBUG uncomment if condition"

This reverts commit 3ea38292497d5072f346f0248244c4c2ec1fc985.

lorentzenchr committed:
TST increase max_depth and n_samples in test_grower_interaction_constraints

lorentzenchr committed:
ENH add default group to interaction constraints

lorentzenchr committed:
TST test_check_interaction_cst

lorentzenchr committed:
DOC udpate docstring of interaction_cst with default group

lorentzenchr commented:
I consider this PR now mature enough to replace WIP by MRG. I am, however, troubled by the out of bound error as reported in https://github.com/scikit-learn/scikit-learn/pull/21020#issuecomment-922337842.

lorentzenchr committed:
DOC add whatsnew

lorentzenchr committed:
DEBUG

ogrisel commented:
I will do a proper review soon-ish, but in the mean time I think this PR would benefit from adding or extending an example to show how this can be used to build non-linear interpretable models without feature interactions, as an alternative to spline-based GAMs (not yet available in scikit-learn) and splinetransformer + linear models.

A partial dependency plot that demonstrate the non-linear marginal effect of each feature on the target while not having any pairwise interactions would be interesting in particular. I am not sure which dataset would be the most suitable for such an example though.

jjerphan reviewed (commented):
This is a first pass on this PR.

I will try to reproduce and diagnose the problem mentioned in https://github.com/scikit-learn/scikit-learn/pull/21020#issuecomment-920375953.

ogrisel commented on sklearn/ensemble/_hist_gradient_boosting/gradient_boosting.py:
```diff
@@ -45,6 +46,7 @@ def __init__(
         max_bins,
         categorical_features,
         monotonic_cst,
+        interaction_cst,
```
I think it's too late to change because we have already released this estimator with the `monotonic_cst` parameter.

lorentzenchr committed:
TST make test_split_interaction_constraints more tighter

lorentzenchr committed:
better comments and less typos

lorentzenchr committed:
Merge branch 'main' into hgbt_interaction_constraints

jjerphan commented on sklearn/ensemble/_hist_gradient_boosting/splitting.pyx:
```diff
@@ -381,18 +388,37 @@ cdef class Splitter:
             # map indices in left/right_indices_buffer back into
             # sample_indices. This also updates self.partition since
             # sample_indices is a view.
+            with gil:
+                print("\nsplit_indices")
+                print(f"n_threads = {n_threads}")
+                print(f"len(sample_indices) = {len(sample_indices)}")
+                print(f"sample_indices = {np.array(sample_indices)}")
             for thread_idx in prange(n_threads, schedule='static',
                                      chunksize=1):
+                with gil:
+                    print(
+                        f"\nthread_idx = {np.array(thread_idx)}"
+                        f"\noffset_in_buffers = {np.array(offset_in_buffers)}"
+                        f"\nleft_counts = {np.array(left_counts)}"
+                        f"\nleft_offset = {np.array(left_offset)}"
+                        f"\nleft_indices_buffer = {np.array(left_indices_buffer)}"
+                        f"\nright_counts = {np.array(right_counts)}"
+                        f"\nright_offset = {np.array(right_offset)}"
+                        f"\nright_indices_buffer = {np.array(right_indices_buffer)}",
+                        flush=True,
+                    )
```
It looks like temporary changes for debugging have been committed.

lorentzenchr replied:
Uups. I was short in time. The "DEBUG" commit is reverted. I hope is fine again.

lorentzenchr committed:
Revert "DEBUG"

This reverts commit a18b5ee5151ac528a600af594d5080fff3be1810.

amueller commented:
I'm a bit surprised that we consider including this but not #19914 when #19914 is a much easier to understand and better evaluated and more widely cited version of this.

mayer79 commented:
@amueller : GAMS via gradient boosting is an excellent modeling technique, but it is already available with the current implementation, using HistGradientBooster with two leave nodes. Basically, there is nothing to do except maybe to write a nice little example.

Interaction constraints are more flexible: e.g. you can fit a model additive in three key features and non-additive in a bunch of other variables (e.g. text embeddings, location variables etc.). Currently, such things cannot be done, so it makes sense to implement it.

amueller commented:
@mayer79 I agree that interaction constraints are more flexible, but GAMs via gradient boosting are not available as they are not usually stumps, and it has been found that cyclic boosting is important (iterating through the features) for interpretability, and that's not available either.
Also, the work I referenced has pairwise interactions.

I agree that it makes sense to implement this, but we don't even have the more basic and well-published method available.

mayer79 commented:
> @mayer79 I agree that interaction constraints are more flexible, but GAMs via gradient boosting are not available as they are not usually stumps (I discussed with one of the researchers and they said the results when restricting to stumps are quite a bit worse), and it has been found that cyclic boosting is important (iterating through the features) for interpretability, and that's not available either. Also, the work I referenced has pairwise interactions.
> 
> I agree that it makes sense to implement more general interaction constraints, but we don't even have the more basic and well-published method available.

EMBs are definitively interesting, no question! (A simple way to emulate an EMB with pairwise interactions: use trees with 3 terminal nodes and a per-tree column sampling rate of 2/nfeatures.)

NicolasHug commented:
> I'm a bit surprised that we consider including this but not #19914 when #19914 is a much easier to understand and better evaluated and more widely cited version of this.

I don't think we're not considering #19914 for inclusion, I think it's just that #19914 is still WIP while this one isn't. It's also more straighforward regarding the implementation.

amueller commented:
Oh I thought it's WIP because at the core dev meeting people were opposed to inclusion? @thomasjpfan can you confirm or deny? ;)

lorentzenchr committed:
Merge branch 'main' into hgbt_interaction_constraints

lorentzenchr committed:
DOC address review comments for docstrings

lorentzenchr committed:
TST reviewer suggestion for improved grower test

lorentzenchr committed:
TST check interaction constraints numerically

lorentzenchr committed:
Merge branch 'main' into hgbt_interaction_constraints

lorentzenchr committed:
EXA add interaction constraints to partial dependence

lorentzenchr committed:
CLN colon in example

lorentzenchr committed:
CLN fix whatsnew

ogrisel reviewed (commented):
Minor remarks in a pending review I started a while ago but did not complete yet:

lorentzenchr committed:
TST better error messages

lorentzenchr committed:
EXA add 1D ice plots to see parallel lines

lorentzenchr committed:
TST rely more on default values

amueller commented on sklearn/ensemble/_hist_gradient_boosting/gradient_boosting.py:
```diff
@@ -1093,6 +1131,22 @@ class HistGradientBoostingRegressor(RegressorMixin, BaseHistGradientBoosting):
 
         .. versionadded:: 0.23
 
+    interaction_cst : iterable of iterables of int, default=None
```
> The resulting trees are not identical, but the resulting model structure is in the sense that the interaction constraints are fulfilled in both cases.

The constraint is fulfilled but the models are not equivalent in any way, right? Aka what @ogrisel  said, it's quite a different model.

> But the tree will usually use three features.

One of the reasons people restrict to pairwise interactions is so that the full model can be visualized. That's much harder with three features. There is no way to achieve trees that are on pairs of features with this PR, right?

amueller commented:
@lorentzenchr that seems like a good way forward to me. What I'm not quite sure about is whether it's possible to create trees that only have pairwise interactions so that partial dependence plots + pairwise plots would show the whole model.

mayer79 commented:
> @lorentzenchr that seems like a good way forward to me. What I'm not quite sure about is whether it's possible to create trees that only have pairwise interactions so that partial dependence plots + pairwise plots would show the whole model.

This is not possible (as with any model using pairwise interactions,  I'd say). The typical use case of interaction constraints will be to fit partly additive models, i.e., models that are additive in one or more features and fully flexible in others. For example you can keep a model additive in "gender". After fitting, the gender effect can easily be removed from the model, leading to an almost discrimination free model.

lorentzenchr commented:
> What I'm not quite sure about is whether it's possible to create trees that only have pairwise interactions so that partial dependence plots + pairwise plots would show the whole model.

@amueller I'd say: **Let's merge and try out**.
As discussed in https://github.com/scikit-learn/scikit-learn/pull/21020#discussion_r735000184, all pairwise interactions can be specified with `interaction_cst = list(itertools.combinations(range(n_features), 2)))`. The problem with this is that it could produce a huge list of O(n_features**2) elements. This could later be mitigated by a user friendly option like `interaction_cst="all_pairs"`, maybe even with a more efficient implementation for this special case. (Unexlusive) Or we additionally pursue #19914.

lorentzenchr committed:
Merge branch 'main' into hgbt_interaction_constraints

lorentzenchr committed:
DOC add blank lines in whats_new

lorentzenchr committed:
DOC remove 1.0.1 entry in whats_new 1.1

amueller commented:
> If one restricts the number of leaves to three, the trees will at most use two features, but that is true even without constraints.

As said multiple times above, this just results in completely different models, so I don't think it's a valid argument?

> Still, as with any model using pairwise interactions, pairwise pdps won't give a fully transparent model description of feature effects.

I don't think that's correct. I think pairwise PDPs will give the full information about the model.

>  interaction_cst = list(itertools.combinations(range(n_features), 2)))

as discussed there, and @mayer79  said, "As such, each tree prediction will use only two features. But the tree will usually use three features." so this is not what I would call a tree with pairwise interactions.

Me and @mayer79  clearly have different use-cases in mind, which is fine, but I would like to be clear which use-cases are supported by the implementation, and it doesn't seem that my definition of pairwise interactions (i.e. the model is an additive model in univariate and pairwise effects) is supported.

amueller commented:
> For example you can keep a model additive in "gender". After fitting, the gender effect can easily be removed from the model, leading to an almost discrimination free model.

This is not true if gender is independent of the other features/covariates in the original data, which is not a good assumption in most cases.

lorentzenchr commented:
Marking as high priority according to https://scikit-learn.fondation-inria.fr/technical-committee-november-5-2020-fr/ and https://scikit-learn.fondation-inria.fr/technical-committee-june-2-2021-fr/.

lesteve committed:
Merge branch 'main' of https://github.com/scikit-learn/scikit-learn into hgbt_interaction_constraints

lorentzenchr committed:
Merge branch 'main' into hgbt_interaction_constraints

jeremiedbb commented:
Let's move to 1.2 since it requires more reviews

lorentzenchr committed:
Merge branch 'main' into hgbt_interaction_constraints

lorentzenchr committed:
CLN fix merge with parameter validation

lorentzenchr commented:
Despite many very welcome review comments, this PR still needs two reviewer approvals. I'm pretty confident that all mentioned concerns on the supported use cases can be resolved eventually. Having this would enable several follow-up PRs for improvements.

lorentzenchr committed:
DOC move whatsnew to v1.2

jjerphan commented:
> Despite many very welcome review comments, this PR still needs two reviewer approvals. I'm pretty confident that all mentioned concerns on the supported use cases can be resolved eventually. Having this would enable several follow-up PRs for improvements.

@lorentzenchr: that's somewhere on my TODO-maxheap. Thanks for your patience.

lorentzenchr commented:
@jjerphan Any ETA of your TODO-maxheap?

thomasjpfan commented on sklearn/ensemble/_hist_gradient_boosting/grower.py:
```diff
@@ -554,6 +613,45 @@ def split_next(self):
+
+        Returns
+        -------
+        allowed_features : None or ndarray, dtype=int
```
```suggestion
        allowed_features :  ndarray, dtype=uint32
```

lorentzenchr replied:
Why remove `None` as valid option?

thomasjpfan commented on sklearn/ensemble/_hist_gradient_boosting/grower.py:
```diff
@@ -554,6 +613,45 @@ def split_next(self):
+        -------
+        allowed_features : None or ndarray, dtype=int
+            Indices of features allowed to split for children.
+        interaction_cst_indices : None or list of ints
```
```suggestion
        interaction_cst_indices : list of ints
```

lorentzenchr replied:
Why remove `None` as valid option?

thomasjpfan replied:
In the current implementation, the return is always a list and never `None`.

lorentzenchr replied:
🙈 That happens if you look into your own code 1 year later 🙈

thomasjpfan commented on sklearn/ensemble/_hist_gradient_boosting/splitting.pyx:
```diff
@@ -136,6 +136,10 @@ cdef class Splitter:
         feature.
     is_categorical : ndarray of bool of shape (n_features,)
         Indicates categorical features.
+    monotonic_cst : ndarray of shape (n_features,), dtype=int
```
Can this be in a separate small PR?

lorentzenchr replied:
See #24191.

thomasjpfan commented on sklearn/ensemble/_hist_gradient_boosting/splitting.pyx:
```diff
@@ -149,6 +153,8 @@ cdef class Splitter:
         be ignored.
     hessians_are_constant: bool, default is False
         Whether hessians are constant.
+    n_threads : int, default=1
```
Can this be in a separate small PR?

thomasjpfan commented on sklearn/ensemble/_hist_gradient_boosting/grower.py:
```diff
@@ -554,6 +613,45 @@ def split_next(self):
+    def _compute_interactions(self, node):
+        r"""Compute features allowed by interactions to be inherited by child nodes.
+
+        Example: Assume constraints [{0, 1}, {1, 2}].
```
I think this information would be great to have in the User Guide:

https://github.com/scikit-learn/scikit-learn/blob/57e35773f708f6d6129d643a7a237b2fbe0779eb/doc/modules/ensemble.rst?plain=1#L911-L914

What do you think is the best reference for this method?

XREF: I read about the method from the [XGBoost docs](https://xgboost.readthedocs.io/en/latest/tutorials/feature_interaction_constraint.html).

lorentzenchr replied:
I added a user guide entry with https://github.com/scikit-learn/scikit-learn/pull/21020/commits/b9d880bf105f454b1a17789b39e6df3d3dcf802e, but I'm not a Sphinx expert :grimacing:. Help wanted.

Regarding a reference: I'm not aware of a good reference in the literature that explains or depicts it with trees, but the same is true for other parameters. LightGBM and XGBoost are the references for me. While XGBoost has by far the better documentation for interaction constraints, it has/had a wrong implementation for overlapping groups (AFAIK they therefore forbid overlapping groups by now).

lorentzenchr commented on sklearn/ensemble/_hist_gradient_boosting/grower.py:
```diff
@@ -182,6 +195,28 @@ class TreeGrower:
+
+    Attributes
+    ----------
+    histogram_builder : HistogramBuilder
```
I opened #24191.

lorentzenchr committed:
CLN move missing docstring additions to other PR

lorentzenchr committed:
DOC add user guide entry

lorentzenchr committed:
MNT change versionadded to 1.2

jjerphan commented:
> @jjerphan Any ETA of your TODO-maxheap?

I need to block days to do a complete review of it — sometimes I have been trying to do lately but that I've not been able to, much to my regret.

lorentzenchr committed:
DOC use code-block:: text

lorentzenchr committed:
DOC allowed_features has dtype uint32

lorentzenchr committed:
DOC remove None from interaction_cst_indices

lorentzenchr committed:
Merge branch 'main' into hgbt_interaction_constraints

lorentzenchr committed:
DOC fix typo

lorentzenchr committed:
DOC try none to switch off language highlightning

lorentzenchr committed:
DOC address features by numbers

mayer79 reviewed (approved):
Very nice. In contrast to XGBoost, also overlapping constraint sets are allowed.

lorentzenchr committed:
Merge branch 'main' into hgbt_interaction_constraints

lorentzenchr commented:
@NicolasHug You did one of the first thorough reviews. Any chance I get an approval from you?

thomasjpfan commented on sklearn/ensemble/_hist_gradient_boosting/grower.py:
```diff
@@ -576,6 +613,45 @@ def split_next(self):
+        #  - This is for nodes that are already split and have a
+        #    node.split_info.feature_idx.
+        allowed_features = set()
+        interaction_cst_indices = list()
```
Nit:

```suggestion
        interaction_cst_indices = []
```

lorentzenchr replied:
Ok. Even if I find the explicit `list()` nicer here just after the `set()`.

thomasjpfan commented on sklearn/ensemble/_hist_gradient_boosting/grower.py:
```diff
@@ -168,6 +180,7 @@ class TreeGrower:
         and 0 respectively correspond to a positive constraint, negative
         constraint and no constraint. Read more in the :ref:`User Guide
         <monotonic_cst_gbdt>`.
+    interaction_cst : list of sets of integers
```
```suggestion
    interaction_cst : list of sets of integers, default=None
        List of interaction constraints.
```

thomasjpfan commented on sklearn/ensemble/_hist_gradient_boosting/grower.py:
```diff
@@ -400,6 +415,14 @@ def _intilialize_root(self, gradients, hessians, hessians_are_constant):
         )
+
+        if self.interaction_cst is not None:
+            self.root.interaction_cst_indices = list(range(len(self.interaction_cst)))
```
From the way we use `interaction_cst_indices`, I think we can get away with not creating the list? Something like this:

```suggestion
            self.root.interaction_cst_indices = range(len(self.interaction_cst))
```

I have no strong opinion on this.

lorentzenchr replied:
Generators, good idea, might help in a follow-up for several things.

thomasjpfan commented on sklearn/ensemble/_hist_gradient_boosting/tests/test_gradient_boosting.py:
```diff
@@ -1082,6 +1123,68 @@ def test_uint8_predict(Est):
+    )
+
+
+# TODO: Remove in v1.2
```
```suggestion
```

thomasjpfan commented on sklearn/ensemble/_hist_gradient_boosting/tests/test_grower.py:
```diff
@@ -567,3 +567,69 @@ def test_ohe_equivalence(min_samples_leaf, n_unique_categories, target):
+                res.extend(get_all_children(n))
+        return res
+
+    for seed in range(20):
```
Is this a candidate for using `global_random_seed`?

lorentzenchr replied:
Not at this place. The loop over 20 seeds is needed to ensure that every feature was split on at least once in `TreeGrower` (which is tested at the end of the test).

With `global_random_seed` instead of the loop, it not possible (or hard) to ensure that.

jjerphan replied:
```suggestion
    # Ensure that every feature was split on at least once in `TreeGrower`
    # on different trials.
    for seed in range(20):
```

thomasjpfan commented on examples/inspection/plot_partial_dependence.py:
```diff
@@ -255,6 +255,85 @@
+print("Training interaction constraint HistGradientBoostingRegressor...")
+tic = time()
+est_no_interactions = HistGradientBoostingRegressor(
+    interaction_cst=list(zip(range(X_train.shape[1])))
```
I think the following is clearer than `list` + `zip` + `range`?

```suggestion
    interaction_cst=[[i] for i in range(X_train.shape[1])]
```

thomasjpfan reviewed (commented):
Here is a review on the implementation. I still need to look at the tests.

lorentzenchr committed:
address reviewer comments

lorentzenchr commented:
> Can we add a note in the User Guide about how our implementation differs from the XGBoost's implementation?

I'd like to, but it seems this is not yet settled for XGBoost, see https://github.com/dmlc/xgboost/issues/7115 and the unmerged https://github.com/dmlc/xgboost/pull/7117.
> Also, does this PR differ from LightGBM's implementation?

See https://github.com/microsoft/LightGBM/issues/4481. I'll add a note in the UG.

lorentzenchr committed:
DOC add note about LightGBM logic

lorentzenchr committed:
Merge branch 'main' into hgbt_interaction_constraints

Vincent-Maladiere commented on doc/modules/ensemble.rst:
```diff
@@ -1169,6 +1169,37 @@ supported for multiclass context.
+
+A priori, the histogram gradient boosting trees are allowed to use any feature
+to split a node into child nodes. This creates so called interactions between
+features, i.e. usage of different features as split along a brang. Sometimes,
```
typo? "branch"

Vincent-Maladiere replied:
```suggestion
features, i.e. usage of different features as split along a branch. Sometimes,
```

lorentzenchr committed:
DOC fix typo

lorentzenchr committed:
Merge branch 'main' into hgbt_interaction_constraints

lorentzenchr committed:
CLN better comment on test construction

lorentzenchr commented:
@thomasjpfan Is there anything I can do (help, support, trade) to get a LGTM from you? You gave the most thorough review, and I think all of your comments are addressed, no loose ends.

thomasjpfan commented on sklearn/ensemble/_hist_gradient_boosting/tests/test_gradient_boosting.py:
```diff
@@ -1083,6 +1124,69 @@ def test_uint8_predict(Est):
+    # y = x0 + x1 + 5 * x0 * x1
+    y = np.c_[X, 5 * X[:, 0] * X[:, 1]].sum(axis=1)
+
+    est = HistGradientBoostingRegressor()
```
So this is stable for our CI and other package managers:

```suggestion
    est = HistGradientBoostingRegressor(random_state=42)
```

thomasjpfan commented on sklearn/ensemble/_hist_gradient_boosting/tests/test_gradient_boosting.py:
```diff
@@ -1083,6 +1124,69 @@ def test_uint8_predict(Est):
+
+    est = HistGradientBoostingRegressor()
+    est.fit(X, y)
+    est_no_interactions = HistGradientBoostingRegressor(interaction_cst=[{0}, {1}])
```
Same here regarding `random_state`.

thomasjpfan commented on sklearn/ensemble/_hist_gradient_boosting/tests/test_gradient_boosting.py:
```diff
@@ -1083,6 +1124,69 @@ def test_uint8_predict(Est):
+        atol=1e-12,
+    )
+
+    # Correct result of the expressions is 5 * delta**2. But this is hard to achieve by
+    # a fitted tree-based model. The expression should, however, at least be positive!
```
I'm wondering if we should include this test here. It's not really checking the interaction constraints.

Also it fails when there are not enough trees:

```python
def test_interaction_cst_numerically_v2():
    """Check that interaction constraints have no forbidden interactions."""
    rng = np.random.RandomState(42)
    n_samples = 1000
    X = rng.uniform(size=(n_samples, 2))
    y = np.hstack((X, 5 * X[:, [0]] * X[:, [1]])).sum(axis=1)

    est = HistGradientBoostingRegressor(max_iter=, random_state=0)
    est.fit(X, y)

    delta = 0.25
    X_test = X[(X[:, 0] < 1 - delta) & (X[:, 1] < 1 - delta)]
    X_delta_0 = X_test + [delta, 0]
    X_delta_1 = X_test + [0, delta]
    X_delta_0_1 = X_test + [delta, delta]
    assert np.all(
        est.predict(X_delta_0_1)
        + est.predict(X_test)
        - est.predict(X_delta_0)
        - est.predict(X_delta_1)
        > 0.01
    )
```

lorentzenchr replied:
That is a bit the point: Tree based models are very bad in capturing 2-dimensional (or higher) structures which are not coordinate/axis aligned (not well approximated with rectangles/hypercubes).

While I do not insist on this test, I find it very neat. It is cheap to compute and needs only these few lines of code if placed here. Semantically, it's not the best place.

thomasjpfan replied:
I see. Okay, let's keep it.

thomasjpfan replied:
From this conversation, what do you think of adding "with 100 iterations"?

```suggestion
    # Correct result of the expressions is 5 * delta**2. But this is hard to achieve by
    # a fitted tree-based model. However, with 100 iterations the expression should
    # at least be positive!
```

thomasjpfan commented on sklearn/ensemble/_hist_gradient_boosting/grower.py:
```diff
@@ -578,6 +616,45 @@ def split_next(self):
+          / \
+         1   2    <- Left split still fulfills both constraint groups.
+        / \ / \      Right split at feature 2 has only group {1, 2} from now on.
+
```
Since this is docstring is for developers:

```suggestion
        LightGBM uses the same logic for overlapping groups. See
        https://github.com/microsoft/LightGBM/issues/4481 for details.
```

thomasjpfan commented on sklearn/ensemble/_hist_gradient_boosting/splitting.pyx:
```diff
@@ -461,38 +465,51 @@ cdef class Splitter:
             int n_threads = self.n_threads
+            bint has_interaction_cst = False
+
+        if allowed_features is not None:
+            has_interaction_cst = True
+            n_allowed_features = allowed_features.shape[0]
```
Nit: I think this makes it clearer about how `n_allowed_features` is defined:

```suggestion
        has_interaction_cst = allowed_features is not None
        if has_interaction_cst:
            n_allowed_features = allowed_features.shape[0]
        else:
            n_allowed_features = self.n_features
```

lorentzenchr commented on examples/inspection/plot_partial_dependence.py:
```diff
@@ -255,6 +255,85 @@
+# %%
+# All 4 plots show parallel lines meaning there is no interaction in the model.
+# (Note that to see the same with a
+# :class:`~sklearn.ensemble.HistGradientBoostingClassifier`, we would need to
```
You're right. The statement would be valid for `HistGradientBoostingRegressor(loss="poisson")`, but I'll remove it from here. This subject is already discussed in #18309.

lorentzenchr committed:
EXA review comments

lorentzenchr commented on sklearn/ensemble/_hist_gradient_boosting/gradient_boosting.py:
```diff
@@ -252,6 +256,42 @@ def _check_categories(self, X):
+        constraints = [set(group) for group in self.interaction_cst]
+
+        # Add all not listed features as own group by default.
+        rest = set(range(n_features)).difference(set().union(*constraints))
```
Me too;-)

lorentzenchr commented on sklearn/ensemble/_hist_gradient_boosting/tests/test_gradient_boosting.py:
```diff
@@ -1083,6 +1124,69 @@ def test_uint8_predict(Est):
 
 
 # TODO(1.3): Remove
+@pytest.mark.parametrize(
```
Well spotted. Thanks.

lorentzenchr commented on sklearn/ensemble/_hist_gradient_boosting/tests/test_gradient_boosting.py:
```diff
@@ -1083,6 +1124,69 @@ def test_uint8_predict(Est):
+    X = rng.uniform(size=(n_samples, 2))
+    # Construct y with a strong interaction term
+    # y = x0 + x1 + 5 * x0 * x1
+    y = np.c_[X, 5 * X[:, 0] * X[:, 1]].sum(axis=1)
```
I easily comply to that preference. Do you know the reasons?

thomasjpfan replied:
I guess not many are familiar with `np.c_`? `np.c_` is [weirdly documented](https://numpy.org/doc/stable/reference/generated/numpy.c_.html) compared to [np.hstack](https://numpy.org/doc/stable/reference/generated/numpy.hstack.html)

lorentzenchr commented on sklearn/ensemble/_hist_gradient_boosting/tests/test_gradient_boosting.py:
```diff
@@ -1083,6 +1124,69 @@ def test_uint8_predict(Est):
+    assert est._check_interaction_cst(n_features) == result
+
+
+def test_interaction_cst_numerically():
```
I can't reproduce your timings locally. The time is dominated by the fit of the 2 models. Both take about the same time.

thomasjpfan replied:
Ah I think there is something wrong with my setup. It is running slower when `OMP_NUM_THREADS!=1`. When I set `OMP_NUM_THREADS=1` everything runs fast.

I'm okay with the runtime and leaving the test as is.

lorentzenchr committed:
ENH improvements from Thomas review comments

jjerphan commented on sklearn/ensemble/_hist_gradient_boosting/gradient_boosting.py:
```diff
@@ -1191,6 +1235,22 @@ class HistGradientBoostingRegressor(RegressorMixin, BaseHistGradientBoosting):
+
+        For instance, with 5 features in total, `interaction_cst=[{0, 1}]`
+        is equivalent to `interaction_cst=[{0, 1}, {2, 3, 4}]`,
+        and specifies that a single branch of a tree will either only split
+        on features 0 and 1 or only split on features 2, 3 and 4.
```
Is my understanding exact?
```suggestion
        and specifies that each single branch of a tree will either be able
        to only split on features 0 and 1 or only split on features 2, 3 and 4.
```

lorentzenchr replied:
I went with
```
specifies that each branch of a tree will either only split
        on features 0 and 1 or only split on features 2, 3 and 4.
```
The "being able to split" seems overcomplicated to me. It's not the ability that counts, but the fact that it does.

jjerphan commented on sklearn/ensemble/_hist_gradient_boosting/gradient_boosting.py:
```diff
@@ -1505,6 +1567,22 @@ class HistGradientBoostingClassifier(ClassifierMixin, BaseHistGradientBoosting):
+
+        For instance, with 5 features in total, `interaction_cst=[{0, 1}]`
+        is equivalent to `interaction_cst=[{0, 1}, {2, 3, 4}]`,
+        and specifies that a single branch of a tree will either only split
+        on features 0 and 1 or only split on features 2, 3 and 4.
```
Is my understanding exact?
```suggestion
        and specifies that each single branch of a tree will either be able
        to only split on features 0 and 1 or only split on features 2, 3 and 4.
```

lorentzenchr replied:
See above.

jjerphan commented on sklearn/ensemble/_hist_gradient_boosting/splitting.pyx:
```diff
@@ -518,12 +537,13 @@ cdef class Splitter:
-                split_infos)
-            split_info = split_infos[best_feature_idx]
+            # split_info is set to the best of split_infos
+            best_split_info_idx = self._find_best_feature_to_split_helper(
+                split_infos, n_allowed_features)
+            split_info = split_infos[best_split_info_idx]
```
Julien's nitpicking at his best.
```suggestion
            best_split_info_idx = self._find_best_feature_to_split_helper(
                split_infos, n_allowed_features,
            )
            split_info = split_infos[best_split_info_idx]
```

lorentzenchr replied:
black for cython!

jjerphan commented on sklearn/ensemble/_hist_gradient_boosting/grower.py:
```diff
@@ -578,6 +616,48 @@ def split_next(self):
+        allowed_features : ndarray, dtype=uint32
+            Indices of features allowed to split for children.
+        interaction_cst_indices : list of ints
+            Indices of the interaction sets that have to be applied on splits of
+            child nodes. The fewer sets the stronger the constraint as fewer sets
+            contain fewer features.
```
I do not understand the last sentence. One can have two sets containing half features each, right? In this case, the tree structure won't be as constrained as their features' interactions' being defined by a lot of singletons? Or is my understanding wrong?

lorentzenchr replied:
It means the choice of possible features to split on is the smaller, the less entries `interaction_cst_indices` has. "Stronger constraint" means having larger impact by only allowing very few features to split on.

jjerphan commented on sklearn/ensemble/_hist_gradient_boosting/tests/test_splitting.py:
```diff
@@ -856,3 +856,98 @@ def test_splitting_categorical_sanity(
+
+    split_features = []
+
+    for i in range(10):
```
```suggestion
    # Ensure that every feature was split on at least once in `TreeGrower`
    # on different trials.
    for i in range(10):
```

jjerphan reviewed (commented):
My (hopefully) penultimate review.

I still need to go through `_compute_interactions` and some tests.

jjerphan commented on sklearn/ensemble/_hist_gradient_boosting/tests/test_grower.py:
```diff
@@ -567,3 +567,69 @@ def test_ohe_equivalence(min_samples_leaf, n_unique_categories, target):
+            # Root's children's allowed_features must be the root's constraints set.
+            assert_array_equal(node.allowed_features, list(root_constraint_set))
+        for node in get_all_children(grower.root):
+            # Nodes accessible from the root must have their index in the root's
+            # constraints set. For example, sets {0, 1} and {1, 2} must not interact
+            # with each other.
+            if not node.is_leaf:
+                assert node.split_info.feature_idx in root_constraint_set
```
I think the comment does not match what is being tested, here.

This is rather:
```suggestion
            # Nodes accessible from the root must have their index in the root's
            # constraints set.
            if not node.is_leaf:
                assert node.split_info.feature_idx in root_constraint_set
```

The other part mentioned (i.e. "For example, sets {0, 1} and {1, 2} must not interact with each other.") must be tested (separately).

I think that the best way to test it is to recursively check that:

```python
   parent_interaction_cst_indices = set(node.interaction_cst_indices.flatten())
   right_interactions_cst_indices = set(
       node.right_child.interaction_cst_indices.flatten()
   )
   left_interactions_cst_indices = set(
       node.left_child.interaction_cst_indices.flatten()
   )

   assert parent_interaction_cst_indices.issupersset(right_interactions_cst_indices)
   assert parent_interaction_cst_indices.issupersset(left_interactions_cst_indices)
```

What do you think?

lorentzenchr replied:
Good catch. You're right. I adapted your version with slight changes.

lorentzenchr replied:
See https://github.com/scikit-learn/scikit-learn/pull/21020/commits/28c45787006fb6803fa62c61601f6ea8948c194f.

jjerphan commented on sklearn/ensemble/_hist_gradient_boosting/tests/test_splitting.py:
```diff
@@ -856,3 +856,98 @@ def test_splitting_categorical_sanity(
+
+    sample_indices = np.arange(n_samples, dtype=np.uint32)
+    all_hessians = np.ones(1, dtype=G_H_DTYPE)
+    sum_hessians = 1 * n_samples
```
```suggestion
    sum_hessians = n_samples
```

lorentzenchr replied:
The happens with copy&paste :smirk:

jjerphan commented on sklearn/ensemble/_hist_gradient_boosting/tests/test_splitting.py:
```diff
@@ -856,3 +856,98 @@ def test_splitting_categorical_sanity(
+            allowed_features=allowed_features,
+        )
+        split_features.append(si_root.feature_idx)
+        assert si_root.feature_idx in {0, 3}
+
+    # make sure feature 0 and feature 3 are split on in the constraint setting
+    assert {0, 3} == set(split_features)
```
With defining:

```
allowed_features_set = set(allowed_features.flatten()) 
```

this can be rewritten as:

```suggestion
        assert si_root.feature_idx in allowed_features_set

    # make sure feature 0 and feature 3 are split on in the constraint setting
    assert allowed_features_set == set(split_features)
```

jjerphan commented on sklearn/ensemble/_hist_gradient_boosting/tests/test_grower.py:
```diff
@@ -567,3 +567,69 @@ def test_ohe_equivalence(min_samples_leaf, n_unique_categories, target):
+            if not node.is_leaf:
+                assert node.split_info.feature_idx in root_constraint_set
+
+    # Make sure that every feature is used at least once as split for the root node.
+    assert len(set(root_feature_splits)) == n_features
```
Is it worth testing this?
```suggestion
    # Make sure that every feature is used at least once as split for the root node.
    assert len(set(root_feature_splits)) == len(set().union(*interaction_cst)) == n_features
```

jjerphan reviewed (approved):
_Alles in :butter:!_ Thank you for your patience, @lorentzenchr.

Here are some last comments.

lorentzenchr commented on examples/inspection/plot_partial_dependence.py:
```diff
@@ -255,6 +255,85 @@
+print("Training interaction constraint HistGradientBoostingRegressor...")
+tic = time()
+est_no_interactions = HistGradientBoostingRegressor(
+    interaction_cst=[[i] for i in range(X_train.shape[1])]
```
This is not the same, isn‘t it?

jjerphan replied:
Ah yes, you're right: I misread.

lorentzenchr commented on examples/inspection/plot_partial_dependence.py:
```diff
@@ -255,6 +255,85 @@
+# Interaction constraints
+# .......................
+#
+# The histogram gradient boosters have an interesting option to constrain
+# possible interactions among features. In the following, we do not allow any
+# interactions and thus render the model as a version of a tree-based boosted
+# generalized additive model (GAM). This makes the model more interpretable
+# as the effect of each feature can be investaged independently of all others.
```
I would prefer not to. The remark on GAM is just a note for the interested reader. The subject is googlable enough.

lorentzenchr commented on examples/inspection/plot_partial_dependence.py:
```diff
@@ -255,6 +255,85 @@
+
+print(f"done in {time() - tic:.3f}s")
+display.figure_.suptitle(
+    "Partial dependence of house value with Gradient Boosting\n"
+    "and no interactions allowed"
```
In the plots above, its also just "Gradient Boosting". I can adapt each an every one, but I think it's fine as it is: HGBT is GB.

lorentzenchr commented on examples/inspection/plot_partial_dependence.py:
```diff
@@ -255,6 +255,85 @@
+)
+print(f"done in {time() - tic:.3f}s")
+display.figure_.suptitle(
+    "Partial dependence of house value with Gradient Boosting\n"
+    "and no interactions allowed"
```
See above.

lorentzenchr commented on sklearn/ensemble/_hist_gradient_boosting/grower.py:
```diff
@@ -578,6 +616,48 @@ def split_next(self):
+         1   2    <- Left split still fulfills both constraint groups.
+        / \ / \      Right split at feature 2 has only group {1, 2} from now on.
+
+        LightGBM uses the same logic for overlapping groups. See
+        https://github.com/microsoft/LightGBM/issues/4481 for details.
+
```
And we don't know how that will change in the future.

lorentzenchr commented on sklearn/ensemble/_hist_gradient_boosting/grower.py:
```diff
@@ -578,6 +616,48 @@ def split_next(self):
+            if node.split_info.feature_idx in self.interaction_cst[i]:
+                interaction_cst_indices.append(i)
+                allowed_features.update(self.interaction_cst[i])
+        return (
+            np.fromiter(allowed_features, dtype=np.uint32, count=len(allowed_features)),
+            interaction_cst_indices,
+        )
```
I reads longer. Therefore, I prefer to keep the shorter version.

lorentzenchr committed:
CLN Julien's review comments

lorentzenchr committed:
TST fix test_grower_interaction_constraints

lorentzenchr commented:
@jjerphan Thank you so much for your thorough review. Et maitenant, laisson faire notre 🧈 ? (Ça ce dit???)😄

lorentzenchr commented on examples/inspection/plot_partial_dependence.py:
```diff
@@ -255,6 +255,85 @@
 # house age, whereas for values less than two there is a strong dependence on
 # age.
 #
+# Interaction constraints
```
The doc build pipeline/CI passes. So yes, not related to this PR.

lorentzenchr commented:
Supposedly one of my last remarks: We could cite 
>M. Mayer, S.C. Bourassa, M. Hoesli & D.F. Scognamiglio. (2022). "Machine Learning Applications to Land and Structure Valuation". Journal of Risk and Financial Management.
https://doi.org/10.3390/jrfm15050193

In Chapter 2.2, they explain interaction constraints nicely with some diagrams of trees, and refer to the paper that originally proposed interaction constraints:
> Lee, Simon C. K., Sheldon Lin, and Katrien Antonio. 2015. Delta Boosting Machine and Its Application in Actuarial Modeling. Sydney: Institute of Actuaries of Australia.

jjerphan commented:
> @jjerphan Thank you so much for your thorough review. 

You're welcome. Thank you for proposing the call, this gave me a reason to go through your PR again.

> Et maitenant, laisson faire notre 🧈 ? (Ça ce dit???)😄

I don't know about this phrase. If I had to pick one, I would say:  « _Emballé, c'est pesé !_ 📪 »

thomasjpfan commented on doc/modules/ensemble.rst:
```diff
@@ -1174,6 +1174,37 @@ Also, monotonic constraints are not supported for multiclass classification.
+
+Note that features not listed in ``interaction_cst`` are automatically
+assigned an interaction group for themselves. With again 3 features, this
+means that ``[{0}]`` is equivalent to ``[{0}, {1, 2}]``.
```
As noted in https://github.com/scikit-learn/scikit-learn/pull/21020#issuecomment-1273712488, I think "Machine Learning Applications to Land and Structure Valuation" Section 2.2 is worth citing here.

thomasjpfan commented on sklearn/ensemble/_hist_gradient_boosting/tests/test_grower.py:
```diff
@@ -567,3 +567,86 @@ def test_ohe_equivalence(min_samples_leaf, n_unique_categories, target):
+
+    def get_all_children(node):
+        res = []
+        if node.is_leaf:
+            return res
```
According to codecov, this is not being used. I think we can remove the `n.is_leaf` in the `for` loop:

```python
def get_all_children(node):
    res = []
    if node.is_leaf:
        return res
    for n in [node.left_child, node.right_child]:
        res.append(n)
        res.extend(get_all_children(n))
    return res
```

thomasjpfan commented on sklearn/ensemble/_hist_gradient_boosting/tests/test_grower.py:
```diff
@@ -567,3 +567,86 @@ def test_ohe_equivalence(min_samples_leaf, n_unique_categories, target):
+            # constraint set.
+            assert node.split_info.feature_idx in root_constraint_set
+
+    # Make sure that every feature is used at least once as split for the root node.
```
Is this happening by chance or by construction?

lorentzenchr replied:
By chance, with a very high chance, by `for seed in range(20)`.

thomasjpfan replied:
If it's by chance, I think the test is not really checking much? At a high level it's checking similar to:

```python
import numpy as np

rng = np.random.default_rng(42)
n_features = 6
root_feature_splits = []

for _ in range(20):
    root_feature_idx = rng.integers(low=0, high=n_features)
    root_feature_splits.append(root_feature_idx)
    
assert len(set(root_feature_splits)) == n_features
```

lorentzenchr replied:
Except that the `root_feature_idx` is learnd/trained by the TreeGrower. It‘s hard to control on which feature it splits the root node. To avoid the situation where it always splits on, say, feature 5, it is asserted that all features are split on. This makes sure that the test is no nonsense.

lorentzenchr committed:
DOC add reference Mayer 2022

- add reference Machine Learning Applications to Land and Structure Valuation
- add arxiv qualifier for G. Louppe

lorentzenchr committed:
CLN remove if node.is_leaf in for loop

lorentzenchr committed:
CLN fix test_grower_interaction_constraints

thomasjpfan reviewed (approved):
LGTM

lorentzenchr commented:
@jjerphan @thomasjpfan Thank you so much for your time, thoughts and effort while reviewing. It improved this PR a lot. This PR is an important one for me because it enables a lot of really cool use cases of high relevance in practice.
